### PR TITLE
Dockerfile - Add support for arm64 build

### PR DESCRIPTION
**Pull request by dpower4**
aarch64 build.

- Updated dockerfile for aarch64 build
- extend cpu stream compilation for neoverse 
- handle onnxruntime-gpu installation
- third party builds filtering based on arch
- disable cuda decode perf build for non x86

### CODE REVIEW

File: dockerfile/cuda12.4-arm.dockerfile
Comment (polarG) [inline dockerfile/cuda12.4-arm.dockerfile:1]: Can this be integrated with the original cuda12.4 file? Docker native support for choosing arch with '--platform aarch64,amd64' options.

File: third_party/Makefile [hunk 1]
```diff
@@ -18,14 +18,22 @@ NUM_MAKE_JOBS ?= $(shell nproc --ignore=2)
...
+ifeq ($(shell uname -m), aarch64)
+all: cuda
+cuda_with_msccl: cuda cuda_msccl
+cuda: common cuda_cutlass cuda_bandwidthTest cuda_nccl_tests cuda_perftest cuda_gpuburn megatron_lm megatron_deepspeed
+cpu: common cpu_perftest
+common: cpu_stream fio
+else
 all: cuda rocm
 cuda_with_msccl: cuda cuda_msccl
 cuda: common cuda_cutlass cuda_bandwidthTest cuda_nccl_tests cuda_perftest gpcnet cuda_gpuburn megatron_lm megatron_deepspeed
```
Comment (polarG) [inline third_party/Makefile:31]: why gpcnet is excluded from aarch64?

File: setup.py [hunk 2]
```diff
@@ -215,8 +215,8 @@ def run(self):
             ],
             'ort': [
                 'onnx>=1.10.2',
-                'onnxruntime-gpu==1.10.0; python_version<"3.10"',
-                'onnxruntime-gpu; python_version>="3.10"',
+                'onnxruntime-gpu==1.10.0; python_version<"3.10" and platform_machine != "aarch64"',
+                'onnxruntime-gpu; python_version>="3.10" and platform_machine != "aarch64"',
```
Comment (abuccts) [inline setup.py:219]: could we check `== "x86_64"` directly? I don't think it supports other archs

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/CMakeLists.txt [hunk 3]
```diff
@@ -4,114 +4,121 @@
...
-    set(NV_FFMPEG_HDRS ${PC_AVCODEC_INCLUDE_DIRS})
-    find_library(AVCODEC_LIBRARY NAMES avcodec
-    HINTS
-    ${PC_AVCODEC_LIBDIR}
-    ${PC_AVCODEC_LIBRARY_DIRS}
+
+ # Check architecture
+ if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|arm64")
+    message(WARNING "Skipping Cuda decode Performance build. This build only supports x86_64 arch.")
+ else()
```
Comment (abuccts) [inline superbench/benchmarks/micro_benchmarks/cuda_decode_performance/CMakeLists.txt:11]: could you fix the indents here? seems there's one extra space at the beginning of each line

File: third_party/Makefile [hunk 4]
```diff
@@ -18,14 +18,22 @@ NUM_MAKE_JOBS ?= $(shell nproc --ignore=2)
 
 .PHONY: all cuda_with_msccl cuda rocm common cuda_cutlass cuda_bandwidthTest cuda_nccl_tests cuda_perftest cuda_msccl rocm_perftest fio rocm_rccl_tests rocm_rocblas rocm_bandwidthTest gpcnet cuda_gpuburn cpu_stream cpu_hpl directx_amf_encoding_latency directx_amd rocm_hipblaslt megatron_lm megatron_deepspeed apex_rocm
 
-# Build all targets.
+# Build targets.
+ifeq ($(shell uname -m), aarch64)
```
Comment (abuccts) [inline third_party/Makefile:22]: maybe we can just put `common` and `directx_amd` targets inside ifeq and put the rest outside?

File: third_party/stream-tests/Makefile [hunk 5]
```diff
@@ -1,23 +1,42 @@
...
+NEO2_OUTPUT= streamNeo2.exe
+
+ARCH := $(shell uname -m)
 
+ifeq ($(ARCH), aarch64)
+CFLAGS = -Ofast -fopenmp -DNTIMES=200
+CC=gcc
+all: NEO2
+else
+CC= /opt/AMD/aocc-compiler-4.0.0/bin/clang
```
Comment (abuccts) [inline third_party/stream-tests/Makefile:21]: pls unify the spaces arounds `=`:
```
CFLAGS = xx
CC=xx
CC= xx
```

File: dockerfile/cuda12.4-arm.dockerfile [hunk 6]
```diff
@@ -0,0 +1,139 @@
+FROM nvcr.io/nvidia/pytorch:24.05-py3
```
[the remark this answers is not in the document]
Comment (abuccts) [reply dockerfile/cuda12.4-arm.dockerfile:1]: you may refer to [this example](https://stackoverflow.com/a/77726353/7630831) on using one unified dockerfile for multiple platforms, basically you will need
```
```

File: superbench/benchmarks/micro_benchmarks/cpu_stream_performance.py [hunk 7]
```diff
@@ -23,7 +23,7 @@ def __init__(self, name, parameters=''):
         super().__init__(name, parameters)
 
         self._bin_name = 'streamZen3.exe'
-        self.__cpu_arch = ['other', 'zen3', 'zen4']
+        self.__cpu_arch = ['other', 'zen3', 'zen4', 'neo2']
```
Comment (cp5555) [inline superbench/benchmarks/micro_benchmarks/cpu_stream_performance.py:26]: Would you please add new feature in document?

Comment (dpower4) [reply superbench/benchmarks/micro_benchmarks/cpu_stream_performance.py:26]: I checked that previously. The current doc is generic w.r.t cpu stream results. no mention of target processors. Is the plan to list them?

File: third_party/Makefile (hunk 1 above)
Comment (dpower4) [reply third_party/Makefile:31]: Thanks for catching. missed that

File: third_party/stream-tests/Makefile (hunk 5 above)
Comment (dpower4) [reply third_party/stream-tests/Makefile]: made changes for consistency across the file.

File: third_party/Makefile (hunk 4 above)
Comment (dpower4) [reply third_party/Makefile]: cleaned it to remove repetition of targets

File: superbench/benchmarks/micro_benchmarks/cuda_decode_performance/CMakeLists.txt (hunk 3 above)
Comment (dpower4) [reply superbench/benchmarks/micro_benchmarks/cuda_decode_performance/CMakeLists.txt]: Done

File: dockerfile/cuda12.4-arm.dockerfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (dpower4) [reply dockerfile/cuda12.4-arm.dockerfile:1]: Will work with @polarG on the multi platform dockerfile.

**Review by polarG — APPROVED** [review]
👍

**Review by cp5555 — APPROVED** [review]

File: .github/workflows/build-image.yml [hunk 8]
```diff
@@ -28,21 +28,25 @@ jobs:
         - name: cuda12.4
           dockerfile: cuda12.4
           tags: superbench/main:cuda12.4
+          platforms: linux/amd64, linux/arm64
```
Comment (abuccts) [inline .github/workflows/build-image.yml:31]:
```suggestion
          platforms: linux/amd64
```
let's skip the arm64 build in CI first, the emulation seems too slow (10x than amd64)..

File: dockerfile/cuda12.4.dockerfile [hunk 9]
```diff
@@ -19,6 +19,10 @@ FROM nvcr.io/nvidia/pytorch:24.03-py3
 LABEL maintainer="SuperBench"
 
 ENV DEBIAN_FRONTEND=noninteractive
+
+ARG TARGETPLATFORM
+ARG TARGETARCH
```
Comment (abuccts) [inline dockerfile/cuda12.4.dockerfile:24]: maybe move after `ARG NUM_MAKE_JOBS` line to reserve cache for apt-get?

File: dockerfile/cuda12.4.dockerfile [hunk 10]
```diff
@@ -80,40 +84,43 @@ RUN mkdir -p /root/.ssh && \
...
-
-# Install AMD BLIS
-RUN cd /tmp && \
+    rm -rf aocc-compiler-4.0.0_1_amd64.deb && \
+    # Install AMD BLIS
     wget https://download.amd.com/developer/eula/blis/blis-4-0/aocl-blis-linux-aocc-4.0.tar.gz && \
     tar xzf aocl-blis-linux-aocc-4.0.tar.gz && \
     mv amd-blis /opt/AMD && \
-    rm -rf aocl-blis-linux-aocc-4.0.tar.gz
+    rm -rf aocl-blis-linux-aocc-4.0.tar.gz; \
```
Comment (abuccts) [inline dockerfile/cuda12.4.dockerfile:120]: no need for `;` here?

File: dockerfile/cuda12.4.dockerfile [hunk 11]
```diff
@@ -80,40 +84,43 @@ RUN mkdir -p /root/.ssh && \
...
+    else \
+    echo "Skipping Intel MLC, AOCC and AMD Bliss installations for non-amd64 architecture: $TARGETARCH"; \
```
Comment (abuccts) [inline dockerfile/cuda12.4.dockerfile:122]: same to keep consistency

**Review by abuccts — APPROVED** [review]

**Review by abuccts — APPROVED** [review]

**Pull request merged**